### PR TITLE
Contributed pixelshuffle op

### PR DESCRIPTION
**Pull request by MaanavD**
Thanks to @justinchuby for a significant amount of help on this :)

### CODE REVIEW

File: onnxscript/function_libs/torch_lib/ops/core.py [hunk 1]
```diff
@@ -6402,17 +6402,29 @@ def aten_pinverse(self: TensorType, rcond: float = 1e-15) -> TensorType:
...
 
 
 def aten_pixel_unshuffle(self: TensorType, downscale_factor: int) -> TensorType:
     """pixel_unshuffle(Tensor self, int downscale_factor) -> Tensor"""
 
-    raise NotImplementedError()
+    # rank = symbolic_helper._get_tensor_rank(self)
+    # if rank is not None and rank != 4:
+    #     return symbolic_helper._unimplemented("pixel_shuffle", "only support 4d input")
+    # return g.op("DepthToSpace", self, blocksize_i=upscale_factor, mode_s="CRD")
```
Comment (justinchuby) [inline onnxscript/function_libs/torch_lib/ops/core.py:6427]: We can revert this change

**Review by justinchuby — APPROVED** [review]
Thank you!

**Review by titaiwangms — APPROVED** [review]
@MaanavD could you address the review and merge it?

File: onnxscript/function_libs/torch_lib/ops/core.py [hunk 2]
```diff
@@ -6402,17 +6402,29 @@ def aten_pinverse(self: TensorType, rcond: float = 1e-15) -> TensorType:
...
+    self_shape = op.Shape(self)
+    batch = self_shape[:-3]
+    C_out = op.Unsqueeze(self_shape[-3], [0])
+    H_out = op.Unsqueeze(self_shape[-2], [0])
+    W_out = op.Unsqueeze(self_shape[-1], [0])
+    # Reshaping input by collapsing all leading dimensions to match ONNX op requirement (4D)
+    reshaped_self = op.Reshape(self, op.Concat(op.Unsqueeze(-1, [0]), C_out, H_out, W_out, axis=0))
+    depth_to_space_output = op.DepthToSpace(reshaped_self, blocksize=upscale_factor, mode="CRD")
+    output_shape = op.Concat(batch, op.Shape(depth_to_space_output)[1:],axis=0)
+    return op.Reshape(depth_to_space_output, output_shape)
```
Comment (gramalingam) [inline onnxscript/function_libs/torch_lib/ops/core.py:6417]: Something looks odd with the above two lines ... they look like a no-op to me ... it changes just the first dim (at position 0) of the shape of the output tensor ... which must be a no-op for a Reshape.

**Comment by muslll** [comment]
Looking forward for this merge, thanks for the work.

File: onnxscript/function_libs/torch_lib/ops/core.py [hunk 3]
```diff
@@ -6402,17 +6402,29 @@ def aten_pinverse(self: TensorType, rcond: float = 1e-15) -> TensorType:
...
-    raise NotImplementedError()
 
 
 def aten_pixel_unshuffle(self: TensorType, downscale_factor: int) -> TensorType:
     """pixel_unshuffle(Tensor self, int downscale_factor) -> Tensor"""
 
-    raise NotImplementedError()
+    # rank = symbolic_helper._get_tensor_rank(self)
+    # if rank is not None and rank != 4:
+    #     return symbolic_helper._unimplemented("pixel_shuffle", "only support 4d input")
```
Comment (justinchuby) [inline onnxscript/function_libs/torch_lib/ops/core.py:6426]:
```suggestion
    raise NotImplementedError()
```

File: onnxscript/function_libs/torch_lib/ops/core.py (hunk 1 above)
Comment (justinchuby) [inline onnxscript/function_libs/torch_lib/ops/core.py:6427]:
```suggestion
```

**Review by justinchuby — APPROVED** [review]

File: onnxscript/function_libs/torch_lib/ops/core.py [hunk 4]
```diff
@@ -6402,11 +6402,20 @@ def aten_pinverse(self: TensorType, rcond: float = 1e-15) -> TensorType:
...
+    self_shape = op.Shape(self)
+    batch = self_shape[:-3]
+    C_out = op.Unsqueeze(self_shape[-3], [0])
+    H_out = op.Unsqueeze(self_shape[-2], [0])
+    W_out = op.Unsqueeze(self_shape[-1], [0])
+    # Reshaping input by collapsing all leading dimensions to match ONNX op requirement (4D)
+    reshaped_self = op.Reshape(self, op.Concat(op.Unsqueeze(-1, [0]), C_out, H_out, W_out, axis=0))
+    depth_to_space_output = op.DepthToSpace(reshaped_self, blocksize=upscale_factor, mode="CRD")
+    output_shape = op.Concat(batch, op.Shape(depth_to_space_output)[1:],axis=0)
+    return op.Reshape(depth_to_space_output, output_shape)
```
Comment (justinchuby) [inline onnxscript/function_libs/torch_lib/ops/core.py:6418]:
```suggestion
```

File: tests/function_libs/torch_lib/ops_test_data.py [hunk 5]
```diff
@@ -1217,6 +1217,16 @@ def _where_input_wrangler(
         matcher=lambda sample: "weight" in sample.kwargs,
         reason="this Aten overload doesn't accept weight as kwargs",
     ),
+    TorchLibOpInfo(
+        "nn.functional.pixel_shuffle",
+        core_ops.aten_pixel_shuffle,
+    ).xfail(
+        dtypes=(torch.int32, torch.int64),
+        reason="fixme: ONNX Runtime does not support int32/64 inputs",
+    ).xfail(
```
Comment (justinchuby) [inline tests/function_libs/torch_lib/ops_test_data.py:1226]:
```suggestion
    )
    .xfail(
        dtypes=(torch.int32, torch.int64),
        reason="fixme: ONNX Runtime does not support int32/64 inputs",
    )
    .xfail(
```

File: onnxscript/function_libs/torch_lib/ops/core.py [hunk 6]
```diff
@@ -6402,11 +6402,19 @@ def aten_pinverse(self: TensorType, rcond: float = 1e-15) -> TensorType:
 
     raise NotImplementedError()
 
-
-def aten_pixel_shuffle(self: TensorType, upscale_factor: int) -> TensorType:
+@torch_op("aten::pixel_shuffle")
```
Comment (justinchuby) [inline onnxscript/function_libs/torch_lib/ops/core.py:6405]:
```suggestion

@torch_op("aten::pixel_shuffle")
```

File: onnxscript/function_libs/torch_lib/ops/core.py [hunk 7]
```diff
@@ -6402,11 +6402,19 @@ def aten_pinverse(self: TensorType, rcond: float = 1e-15) -> TensorType:
...
-    raise NotImplementedError()
+    self_shape = op.Shape(self)
+    batch = self_shape[:-3]
+    C_out = op.Unsqueeze(self_shape[-3], [0])
+    H_out = op.Unsqueeze(self_shape[-2], [0])
+    W_out = op.Unsqueeze(self_shape[-1], [0])
+    # Reshaping input by collapsing all leading dimensions to match ONNX op requirement (4D)
+    reshaped_self = op.Reshape(self, op.Concat(op.Unsqueeze(-1, [0]), C_out, H_out, W_out, axis=0))
+    depth_to_space_output = op.DepthToSpace(reshaped_self, blocksize=upscale_factor, mode="CRD")
+    output_shape = op.Concat(batch, op.Shape(depth_to_space_output)[1:],axis=0)
```
Comment (justinchuby) [inline onnxscript/function_libs/torch_lib/ops/core.py:6416]:
```suggestion
    reshaped_self = op.Reshape(
        self, op.Concat(op.Unsqueeze(-1, [0]), C_out, H_out, W_out, axis=0)
        )
    depth_to_space_output = op.DepthToSpace(
        reshaped_self, blocksize=upscale_factor, mode="CRD"
        )
    output_shape = op.Concat(batch, op.Shape(depth_to_space_output)[1:], axis=0)
```

File: onnxscript/function_libs/torch_lib/ops/core.py [hunk 8]
```diff
@@ -6403,10 +6403,23 @@ def aten_pinverse(self: TensorType, rcond: float = 1e-15) -> TensorType:
...
+    C_out = op.Unsqueeze(self_shape[-3], [0])
+    H_out = op.Unsqueeze(self_shape[-2], [0])
+    W_out = op.Unsqueeze(self_shape[-1], [0])
+    # Reshaping input by collapsing all leading dimensions to match ONNX op requirement (4D)
+    reshaped_self = op.Reshape(
+        self, op.Concat(op.Unsqueeze(-1, [0]), C_out, H_out, W_out, axis=0)
+        )
+    depth_to_space_output = op.DepthToSpace(
+        reshaped_self, blocksize=upscale_factor, mode="CRD"
+        )
```
Comment (justinchuby) [inline onnxscript/function_libs/torch_lib/ops/core.py:6420]:
```suggestion
    )
```

File: onnxscript/function_libs/torch_lib/ops/core.py [hunk 9]
```diff
@@ -6403,10 +6403,23 @@ def aten_pinverse(self: TensorType, rcond: float = 1e-15) -> TensorType:
...
-    raise NotImplementedError()
+    self_shape = op.Shape(self)
+    batch = self_shape[:-3]
+    C_out = op.Unsqueeze(self_shape[-3], [0])
+    H_out = op.Unsqueeze(self_shape[-2], [0])
+    W_out = op.Unsqueeze(self_shape[-1], [0])
+    # Reshaping input by collapsing all leading dimensions to match ONNX op requirement (4D)
+    reshaped_self = op.Reshape(
+        self, op.Concat(op.Unsqueeze(-1, [0]), C_out, H_out, W_out, axis=0)
+        )
```
Comment (justinchuby) [inline onnxscript/function_libs/torch_lib/ops/core.py:6417]:
```suggestion
    )
```

File: onnxscript/function_libs/torch_lib/ops/core.py [hunk 10]
```diff
@@ -6403,10 +6403,23 @@ def aten_pinverse(self: TensorType, rcond: float = 1e-15) -> TensorType:
...
+def aten_pixel_shuffle(self: TReal, upscale_factor: int) -> TReal:
     """pixel_shuffle(Tensor self, int upscale_factor) -> Tensor"""
-
-    raise NotImplementedError()
+    self_shape = op.Shape(self)
+    batch = self_shape[:-3]
+    C_out = op.Unsqueeze(self_shape[-3], [0])
+    H_out = op.Unsqueeze(self_shape[-2], [0])
+    W_out = op.Unsqueeze(self_shape[-1], [0])
+    # Reshaping input by collapsing all leading dimensions to match ONNX op requirement (4D)
```
Comment (gramalingam) [inline onnxscript/function_libs/torch_lib/ops/core.py:6414]: This is a common operation, but ONNX doesn't give us a convenient way of expressing this. We could potentially extend either the [Flatten op](https://onnx.ai/onnx/operators/onnx__Flatten.html#l-onnx-doc-flatten) or [Reshape op](https://onnx.ai/onnx/operators/onnx__Reshape.html) to support this. This is a one-sided flatten. Alternatively, if reshape has an attribute to align the shape with existing shape starting from the end (instead of beginning), we could specify the new target shape as `[-1, 0, 0, 0]`. Just some idle thoughts ...

Comment (justinchuby) [reply onnxscript/function_libs/torch_lib/ops/core.py:6414]: https://github.com/microsoft/onnxscript/issues/1539
